### ISSUE TITLE
Feature Request: Ability to run a shell command inside flow

### ISSUE THREAD

**Issue by VisenDev** [neurocyte/flow#67]
Similar to `:!` in vim or `alt-!` in emacs

**Comment by neurocyte** [comment]
I usually do this sort of thing with my terminal emulator or terminal multiplexer. I'm not sure what practical value it has these days when nearly every terminal has splits and/or tabs and key bindings to run commands. If you want something to drive a repl I think perhaps a plugin of some sort might be a better way to go.

**Comment by VisenDev** [comment]
perhaps, I usually only use a single terminal pane so sometimes it is convenient to be able to quickly run a command from my editor

**Comment by neurocyte** [comment]
Would you want to have the command output sent to an editor buffer? Or have flow suspend itself and hand over the tty? Flow doesn't support any kind of job control atm, so this could be a lot more complicated than it might seem. Stdio should be pretty simple though.

**Comment by neurocyte** [comment]
The shell command should include a way to send the current selection (or word-at-cursor) to the shell command and also a way to specify that the output should be sent to a new buffer (and allow specifying the language). See #88 for a possible use case.

**Comment by VisenDev** [comment]
> The shell command should include a way to send the current selection (or word-at-cursor) to the shell command and also a way to specify that the output should be sent to a new buffer (and allow specifying the language). See #88 for a possible use case.

Expanding on this - if it would be possible to keep the shell command running in the background and periodically send that shell command new input - that would also more or less address #68

**Comment by neurocyte** [comment]
I've added the ability to bind key to shell commands. The output will be placed in the current buffer. Bin `create_scratch_buffer` first to put the output in a scratch buffer. See `flow.json` for examples.

The second part will be a mini-mode or palette to input commands to be run. Possibly with history fuzzy search.

**Comment by neurocyte** [comment]
The mini-mode and palette are now added. Still working on a few minor bugs.

**Comment by neurocyte** [comment]
closed with 7bcf0e1b758c3540fe0dd342d63e0cb6efb7a2f6

**Comment by edwloef** [comment]
> The shell command should include a way to send the current selection (or word-at-cursor) to the shell command

Was this bit implemented? I can't find any example for it in the `flow.json` file.

**Comment by neurocyte** [comment]
> > The shell command should include a way to send the current selection (or word-at-cursor) to the shell command
> 
> Was this bit implemented? I can't find any example for it in the `flow.json` file.

No, not really. `filter` has this capability, but there is not currently any way to interactively run `filter` commands.

Did you expect the selection to be passed via a command line argument? Or via stdin?

**Comment by edwloef** [comment]
> Did you expect the selection to be passed via a command line argument? Or via stdin?

Honestly, not sure what would be best. Probably the former, so you can do `["shell_execute_insert", "some_command", "{{primary_cursor_position}}"]` or something similar. Writing a little script that turns the cli arguments into stdin arguments shouldn't be hard if you need it, I think. That would open the door to adding more info, such as the currently open file, which I can see being useful.

**Comment by neurocyte** [comment]
Ok, I think this needs a bit more description. I'll close this again here and open a new feature.
